File: server/internal/core/knowledge.go
package core

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// KnowledgeFile serializes flat (filePath/title/tags/createdAt/updatedAt/content),
// matching the desktop Rust `#[serde(flatten)]` wire format.
type KnowledgeFile struct {
	FilePath  string    `json:"filePath"`
	Title     string    `json:"title"`
	Tags      []string  `json:"tags"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
	Content   string    `json:"content"`
}

// meta returns the nested representation used by clients that expect file.meta.*.
func (k KnowledgeFile) MarshalJSON() ([]byte, error) {
	type flat struct {
		FilePath  string    `json:"filePath"`
		Title     string    `json:"title"`
		Tags      []string  `json:"tags"`
		CreatedAt time.Time `json:"createdAt"`
		UpdatedAt time.Time `json:"updatedAt"`
		Content   string    `json:"content"`
		Meta      knowledgeMeta `json:"meta"`
	}
	m := knowledgeMeta{FilePath: k.FilePath, Title: k.Title, Tags: k.Tags, CreatedAt: k.CreatedAt, UpdatedAt: k.UpdatedAt}
	return json.Marshal(flat{k.FilePath, k.Title, k.Tags, k.CreatedAt, k.UpdatedAt, k.Content, m})
}

type knowledgeMeta struct {
	FilePath  string    `json:"filePath"`
	Title     string    `json:"title"`
	Tags      []string  `json:"tags"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

type KnowledgeTreeNode struct {
	Name     string              `json:"name"`
	Path     string              `json:"path"`
	IsDir    bool                `json:"isDir"`
	Children []KnowledgeTreeNode `json:"children"`
}

type CreateKnowledgeInput struct {
	Folder  string    `json:"folder"`
	Title   string    `json:"title"`
	Content *string   `json:"content,omitempty"`
	Tags    *[]string `json:"tags,omitempty"`
}

type UpdateKnowledgeInput struct {
	Content string    `json:"content"`
	Title   *string   `json:"title,omitempty"`
	Tags    *[]string `json:"tags,omitempty"`
}

type knowledgeService struct {
	db   *DB
	root string
}

func newKnowledgeService(db *DB, root string) (*knowledgeService, error) {
	if err := os.MkdirAll(root, 0o755); err != nil {
		return nil, err
	}
	entries, _ := os.ReadDir(root)
	hasFolder := false
	for _, e := range entries {
		if e.IsDir() {
			hasFolder = true
			break
		}
	}
	if !hasFolder {
		for _, f := range []string{"cpp", "graphics", "android", "linux", "ai", "work", "life", "reading"} {
			os.MkdirAll(filepath.Join(root, f), 0o755)
		}
	}
	return &knowledgeService{db: db, root: root}, nil
}

func (s *knowledgeService) listFolders() ([]string, error) {
	entries, err := os.ReadDir(s.root)
	if err != nil {
		return nil, err
	}
	out := []string{}
	for _, e := range entries {
		if e.IsDir() {
			out = append(out, e.Name())
		}
	}
	sort.Strings(out)
	return out, nil
}

func (s *knowledgeService) createFolder(name string) (string, error) {
	folder, err := sanitizeFolderName(name)
	if err != nil {
		return "", err
	}
	p := filepath.Join(s.root, folder)
	if _, err := os.Stat(p); err == nil {
		return "", errf("分类已存在：%s", folder)
	}
	if err := os.MkdirAll(p, 0o755); err != nil {
		return "", err
	}
	return folder, nil
}

func (s *knowledgeService) renameFolder(from, to string) (string, error) {
	ff, err := sanitizeFolderName(from)
	if err != nil {
		return "", err
	}
	tt, err := sanitizeFolderName(to)
	if err != nil {
		return "", err
	}
	if ff == tt {
		return tt, nil
	}
	oldPath := filepath.Join(s.root, ff)
	newPath := filepath.Join(s.root, tt)
	if fi, err := os.Stat(oldPath); err != nil || !fi.IsDir() {
		return "", notFound("分类不存在：" + ff)
	}
	if _, err := os.Stat(newPath); err == nil {
		return "", errf("分类已存在：%s", tt)
	}
	if err := os.Rename(oldPath, newPath); err != nil {
		return "", err
	}
	// Re-key metadata rows under the new folder prefix.
	rows, _ := s.db.sql.Query(`SELECT file_path, title, created_at, updated_at FROM knowledge_meta WHERE file_path LIKE ?`, ff+"/%")
	type row struct{ oldRel, title, created, updated string }
	var items []row
	for rows.Next() {
		var r row
		rows.Scan(&r.oldRel, &r.title, &r.created, &r.updated)
		items = append(items, r)
	}
	rows.Close()
	for _, r := range items {
		newRel := strings.Replace(r.oldRel, ff+"/", tt+"/", 1)
		tags, _ := getEntityTags(s.db.sql, "knowledge_tags", "file_path", r.oldRel)
		content, _ := os.ReadFile(filepath.Join(s.root, newRel))
		body := stripFrontmatter(string(content))
		tx, _ := s.db.sql.Begin()
		tx.Exec(`DELETE FROM knowledge_meta WHERE file_path = ?`, r.oldRel)
		removeSearchIndex(tx, "knowledge", r.oldRel)
		setEntityTags(tx, "knowledge_tags", "file_path", r.oldRel, []string{})
		tx.Exec(`INSERT INTO knowledge_meta (file_path, title, created_at, updated_at) VALUES (?, ?, ?, ?)`, newRel, r.title, r.created, r.updated)
		upsertSearchIndex(tx, "knowledge", newRel, r.title, body)
		if len(tags) > 0 {
			setEntityTags(tx, "knowledge_tags", "file_path", newRel, tags)
		}
		tx.Commit()
	}
	return tt, nil
}

func (s *knowledgeService) deleteFolder(name string) error {
	folder, err := sanitizeFolderName(name)
	if err != nil {
		return err
	}
	p := filepath.Join(s.root, folder)
	if fi, err := os.Stat(p); err != nil || !fi.IsDir() {
		return notFound("分类不存在：" + folder)
	}
	rows, _ := s.db.sql.Query(`SELECT file_path FROM knowledge_meta WHERE file_path LIKE ?`, folder+"/%")
	var files []string
	for rows.Next() {
		var f string
		rows.Scan(&f)
		files = append(files, f)
	}
	rows.Close()
	for _, f := range files {
		s.delete(f)
	}
	return os.RemoveAll(p)
}

func (s *knowledgeService) tree() (KnowledgeTreeNode, error) {
	return buildTreeNode(s.root, s.root), nil
}

func (s *knowledgeService) resolvePath(rel string) (string, error) {
	if filepath.IsAbs(rel) || strings.Contains(rel, "..") {
		return "", errf("invalid path")
	}
	return filepath.Join(s.root, filepath.FromSlash(rel)), nil
}

func (s *knowledgeService) read(rel string) (KnowledgeFile, error) {
	p, err := s.resolvePath(rel)
	if err != nil {
		return KnowledgeFile{}, err
	}
	content, err := os.ReadFile(p)
	if err != nil {
		return KnowledgeFile{}, notFound(rel)
	}
	meta := s.getOrInferMeta(rel, string(content))
	return KnowledgeFile{FilePath: meta.FilePath, Title: meta.Title, Tags: meta.Tags, CreatedAt: meta.CreatedAt, UpdatedAt: meta.UpdatedAt, Content: string(content)}, nil
}

func (s *knowledgeService) create(in CreateKnowledgeInput) (KnowledgeFile, error) {
	folder := strings.Trim(strings.TrimSpace(in.Folder), "/")
	filename := sanitizeFilename(in.Title) + ".md"
	rel := folder + "/" + filename
	p, err := s.resolvePath(rel)
	if err != nil {
		return KnowledgeFile{}, err
	}
	if _, err := os.Stat(p); err == nil {
		return KnowledgeFile{}, errf("file already exists: %s", rel)
	}
	now := nowUTC()
	tags := []string{}
	if in.Tags != nil {
		tags = *in.Tags
	}
	body := "# " + in.Title + "\n"
	if in.Content != nil {
		body = *in.Content
	}
	fileContent := buildMarkdown(in.Title, tags, now, body)
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return KnowledgeFile{}, err
	}
	if err := os.WriteFile(p, []byte(fileContent), 0o644); err != nil {
		return KnowledgeFile{}, err
	}
	tx, err := s.db.sql.Begin()
	if err != nil {
		return KnowledgeFile{}, err
	}
	defer tx.Rollback()
	tx.Exec(`INSERT INTO knowledge_meta (file_path, title, created_at, updated_at) VALUES (?, ?, ?, ?)`, rel, in.Title, rfc3339(now), rfc3339(now))
	setEntityTags(tx, "knowledge_tags", "file_path", rel, tags)
	upsertSearchIndex(tx, "knowledge", rel, in.Title, body)
	if err := tx.Commit(); err != nil {
		return KnowledgeFile{}, err
	}
	return s.read(rel)
}

func (s *knowledgeService) update(rel string, in UpdateKnowledgeInput) (KnowledgeFile, error) {
	p, err := s.resolvePath(rel)
	if err != nil {
		return KnowledgeFile{}, err
	}
	if _, err := os.Stat(p); err != nil {
		return KnowledgeFile{}, notFound(rel)
	}
	existing, err := s.read(rel)
	if err != nil {
		return KnowledgeFile{}, err
	}
	title := existing.Title
	if in.Title != nil {
		title = *in.Title
	}
	tags := existing.Tags
	if in.Tags != nil {
		tags = *in.Tags
	}
	now := nowUTC()
	fileContent := buildMarkdown(title, tags, existing.CreatedAt, in.Content)
	if err := os.WriteFile(p, []byte(fileContent), 0o644); err != nil {
		return KnowledgeFile{}, err
	}
	tx, err := s.db.sql.Begin()
	if err != nil {
		return KnowledgeFile{}, err
	}
	defer tx.Rollback()
	tx.Exec(`INSERT INTO knowledge_meta (file_path, title, created_at, updated_at) VALUES (?, ?, ?, ?)
	         ON CONFLICT(file_path) DO UPDATE SET title=?, updated_at=?`,
		rel, title, rfc3339(existing.CreatedAt), rfc3339(now), title, rfc3339(now))
	setEntityTags(tx, "knowledge_tags", "file_path", rel, tags)
	upsertSearchIndex(tx, "knowledge", rel, title, in.Content)
	if err := tx.Commit(); err != nil {
		return KnowledgeFile{}, err
	}
	return s.read(rel)
}

func (s *knowledgeService) delete(rel string) error {
	p, err := s.resolvePath(rel)
	if err != nil {
		return err
	}
	if _, err := os.Stat(p); err == nil {
		os.Remove(p)
	}
	tx, err := s.db.sql.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	tx.Exec(`DELETE FROM knowledge_meta WHERE file_path = ?`, rel)
	removeSearchIndex(tx, "knowledge", rel)
	return tx.Commit()
}

func (s *knowledgeService) rename(rel, newTitle string) (KnowledgeFile, error) {
	old, err := s.read(rel)
	if err != nil {
		return KnowledgeFile{}, err
	}
	folder := filepath.ToSlash(filepath.Dir(rel))
	if folder == "." {
		folder = ""
	}
	newFilename := sanitizeFilename(newTitle) + ".md"
	newRel := newFilename
	if folder != "" {
		newRel = folder + "/" + newFilename
	}
	if rel != newRel {
		oldPath, _ := s.resolvePath(rel)
		newPath, _ := s.resolvePath(newRel)
		if _, err := os.Stat(newPath); err == nil {
			return KnowledgeFile{}, errf("target file exists")
		}
		if err := os.Rename(oldPath, newPath); err != nil {
			return KnowledgeFile{}, err
		}
		tx, _ := s.db.sql.Begin()
		tx.Exec(`DELETE FROM knowledge_meta WHERE file_path = ?`, rel)
		removeSearchIndex(tx, "knowledge", rel)
		tx.Commit()
	}
	return s.update(newRel, UpdateKnowledgeInput{Content: old.Content, Title: &newTitle, Tags: &old.Tags})
}

func (s *knowledgeService) getOrInferMeta(rel, content string) knowledgeMeta {
	if m, ok := s.getMeta(rel); ok {
		return m
	}
	now := nowUTC()
	return knowledgeMeta{FilePath: rel, Title: inferTitle(rel, content), Tags: []string{}, CreatedAt: now, UpdatedAt: now}
}

func (s *knowledgeService) getMeta(rel string) (knowledgeMeta, bool) {
	var title, created, updated string
	err := s.db.sql.QueryRow(`SELECT file_path, title, created_at, updated_at FROM knowledge_meta WHERE file_path = ?`, rel).
		Scan(&rel, &title, &created, &updated)
	if err != nil {
		return knowledgeMeta{}, false
	}
	tags, _ := getEntityTags(s.db.sql, "knowledge_tags", "file_path", rel)
	if tags == nil {
		tags = []string{}
	}
	return knowledgeMeta{FilePath: rel, Title: title, Tags: tags, CreatedAt: parseRFC3339(created), UpdatedAt: parseRFC3339(updated)}, true
}

func buildTreeNode(root, current string) KnowledgeTreeNode {
	name := filepath.Base(current)
	if current == root {
		name = "knowledge"
	}
	rel, _ := filepath.Rel(root, current)
	rel = filepath.ToSlash(rel)
	if rel == "." {
		rel = ""
	}
	children := []KnowledgeTreeNode{}
	entries, _ := os.ReadDir(current)
	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })
	for _, e := range entries {
		p := filepath.Join(current, e.Name())
		if e.IsDir() {
			children = append(children, buildTreeNode(root, p))
		} else if strings.HasSuffix(e.Name(), ".md") {
			childRel, _ := filepath.Rel(root, p)
			children = append(children, KnowledgeTreeNode{Name: e.Name(), Path: filepath.ToSlash(childRel), IsDir: false, Children: []KnowledgeTreeNode{}})
		}
	}
	return KnowledgeTreeNode{Name: name, Path: rel, IsDir: true, Children: children}
}

func buildMarkdown(title string, tags []string, created time.Time, body string) string {
	tagsYaml := "  []"
	if len(tags) > 0 {
		var b strings.Builder
		for i, t := range tags {
			if i > 0 {
				b.WriteString("\n")
			}
			b.WriteString("  - " + t)
		}
		tagsYaml = b.String()
	}
	now := nowUTC()
	return "---\ntitle: " + title + "\ntags:\n" + tagsYaml +
		"\ncreated: " + created.Format(dateFmt) + "\nupdated: " + now.Format(dateFmt) + "\n---\n\n" + body
}

func inferTitle(rel, content string) string {
	for _, line := range strings.Split(content, "\n") {
		if strings.HasPrefix(line, "# ") {
			return strings.TrimPrefix(line, "# ")
		}
	}
	base := filepath.Base(rel)
	return strings.TrimSuffix(base, filepath.Ext(base))
}

func sanitizeFolderName(s string) (string, error) {
	name := strings.Trim(strings.TrimSpace(s), "/")
	if name == "" {
		return "", errf("分类名不能为空")
	}
	if strings.ContainsAny(name, "/\\") || name == "." || name == ".." {
		return "", errf("分类名不合法")
	}
	if strings.ContainsAny(name, "<>:\"|?*") {
		return "", errf("分类名包含非法字符")
	}
	return name, nil
}

func stripFrontmatter(content string) string {
	trimmed := strings.TrimLeft(content, " \t\r\n")
	if !strings.HasPrefix(trimmed, "---") {
		return content
	}
	rest := trimmed[3:]
	if idx := strings.Index(rest, "\n---"); idx >= 0 {
		return strings.TrimLeft(rest[idx+4:], "\n")
	}
	return content
}

func sanitizeFilename(s string) string {
	var b strings.Builder
	for _, c := range s {
		if strings.ContainsRune("<>:\"/\\|?*", c) {
			b.WriteRune('_')
		} else if c == ' ' {
			b.WriteRune('_')
		} else {
			b.WriteRune(c)
		}
	}
	return b.String()
}

File: src-tauri/src/services/knowledge.rs
use crate::database::{get_entity_tags, remove_search_index, set_entity_tags, upsert_search_index, Database};
use crate::error::{AppError, AppResult};
use crate::models::knowledge::{
    CreateKnowledgeInput, KnowledgeFile, KnowledgeMeta, KnowledgeTreeNode, UpdateKnowledgeInput,
};
use chrono::{DateTime, Utc};
use rusqlite::params;
use std::fs;
use std::path::{Path, PathBuf};

pub struct KnowledgeService<'a> {
    db: &'a Database,
    root: PathBuf,
}

impl<'a> KnowledgeService<'a> {
    pub fn new(db: &'a Database, root: PathBuf) -> AppResult<Self> {
        fs::create_dir_all(&root)?;
        for folder in ["cpp", "graphics", "android", "linux", "ai", "work", "life", "reading"] {
            fs::create_dir_all(root.join(folder))?;
        }
        Ok(Self { db, root })
    }

    pub fn tree(&self) -> AppResult<KnowledgeTreeNode> {
        Ok(build_tree_node(&self.root, &self.root))
    }

    pub fn read(&self, rel_path: &str) -> AppResult<KnowledgeFile> {
        let path = self.resolve_path(rel_path)?;
        if !path.exists() {
            return Err(AppError::NotFound(rel_path.to_string()));
        }
        let content = fs::read_to_string(&path)?;
        let meta = self.get_or_infer_meta(rel_path, &content)?;
        Ok(KnowledgeFile { meta, content })
    }

    pub fn create(&self, input: CreateKnowledgeInput) -> AppResult<KnowledgeFile> {
        let folder = sanitize_segment(&input.folder);
        let filename = format!("{}.md", sanitize_filename(&input.title));
        let rel_path = format!("{folder}/{filename}");
        let path = self.resolve_path(&rel_path)?;

        if path.exists() {
            return Err(AppError::Other(format!("file already exists: {rel_path}")));
        }

        let now = Utc::now();
        let tags = input.tags.unwrap_or_default();
        let body = input.content.unwrap_or_else(|| format!("# {}\n", input.title));
        let file_content = build_markdown(&input.title, &tags, &now, &body);

        fs::create_dir_all(path.parent().unwrap())?;
        fs::write(&path, &file_content)?;

        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute(
                "INSERT INTO knowledge_meta (file_path, title, created_at, updated_at)
                 VALUES (?1, ?2, ?3, ?4)",
                params![rel_path, input.title, now.to_rfc3339(), now.to_rfc3339()],
            )?;
            set_entity_tags(&tx, "knowledge_tags", "file_path", &rel_path, &tags)?;
            upsert_search_index(&tx, "knowledge", &rel_path, &input.title, &body)?;
            tx.commit()?;
            Ok(())
        })?;

        self.read(&rel_path)
    }

    pub fn update(&self, rel_path: &str, input: UpdateKnowledgeInput) -> AppResult<KnowledgeFile> {
        let path = self.resolve_path(rel_path)?;
        if !path.exists() {
            return Err(AppError::NotFound(rel_path.to_string()));
        }

        let existing = self.read(rel_path)?;
        let title = input.title.unwrap_or(existing.meta.title);
        let tags = input.tags.unwrap_or(existing.meta.tags);
        let now = Utc::now();
        let file_content = build_markdown(&title, &tags, &existing.meta.created_at, &input.content);

        fs::write(&path, &file_content)?;

        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute(
                "INSERT INTO knowledge_meta (file_path, title, created_at, updated_at)
                 VALUES (?1, ?2, ?3, ?4)
                 ON CONFLICT(file_path) DO UPDATE SET title=?2, updated_at=?4",
                params![
                    rel_path,
                    title,
                    existing.meta.created_at.to_rfc3339(),
                    now.to_rfc3339(),
                ],
            )?;
            set_entity_tags(&tx, "knowledge_tags", "file_path", rel_path, &tags)?;
            upsert_search_index(&tx, "knowledge", rel_path, &title, &input.content)?;
            tx.commit()?;
            Ok(())
        })?;

        self.read(rel_path)
    }

    pub fn delete(&self, rel_path: &str) -> AppResult<()> {
        let path = self.resolve_path(rel_path)?;
        if path.exists() {
            fs::remove_file(&path)?;
        }
        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute("DELETE FROM knowledge_meta WHERE file_path = ?1", params![rel_path])?;
            remove_search_index(&tx, "knowledge", rel_path)?;
            tx.commit()?;
            Ok(())
        })
    }

    pub fn rename(&self, rel_path: &str, new_title: &str) -> AppResult<KnowledgeFile> {
        let old = self.read(rel_path)?;
        let folder = Path::new(rel_path)
            .parent()
            .and_then(|p| p.to_str())
            .unwrap_or("");
        let new_filename = format!("{}.md", sanitize_filename(new_title));
        let new_rel = if folder.is_empty() {
            new_filename.clone()
        } else {
            format!("{folder}/{new_filename}")
        };

        if rel_path != new_rel {
            let old_path = self.resolve_path(rel_path)?;
            let new_path = self.resolve_path(&new_rel)?;
            if new_path.exists() {
                return Err(AppError::Other("target file exists".into()));
            }
            fs::rename(&old_path, &new_path)?;
            self.db.with_conn(|conn| {
                conn.execute("DELETE FROM knowledge_meta WHERE file_path = ?1", params![rel_path])?;
                remove_search_index(conn, "knowledge", rel_path)?;
                Ok(())
            })?;
        }

        self.update(
            &new_rel,
            UpdateKnowledgeInput {
                content: old.content,
                title: Some(new_title.to_string()),
                tags: Some(old.meta.tags),
            },
        )
    }

    /// Full rebuild of knowledge search index (recovery / rare repair).
    #[allow(dead_code)]
    pub fn reindex_all(&self) -> AppResult<()> {
        for entry in walkdir::WalkDir::new(&self.root)
            .into_iter()
            .filter_map(|e| e.ok())
            .filter(|e| e.path().extension().map(|x| x == "md").unwrap_or(false))
        {
            let path = entry.path();
            let rel = path
                .strip_prefix(&self.root)
                .unwrap()
                .to_string_lossy()
                .replace('\\', "/");
            let content = fs::read_to_string(path)?;
            let meta = self.get_or_infer_meta(&rel, &content)?;
            self.db.with_conn(|conn| {
                conn.execute(
                    "INSERT INTO knowledge_meta (file_path, title, created_at, updated_at)
                     VALUES (?1, ?2, ?3, ?4)
                     ON CONFLICT(file_path) DO UPDATE SET title=?2, updated_at=?4",
                    params![
                        rel,
                        meta.title,
                        meta.created_at.to_rfc3339(),
                        meta.updated_at.to_rfc3339(),
                    ],
                )?;
                upsert_search_index(&conn, "knowledge", &rel, &meta.title, &content)?;
                Ok(())
            })?;
        }
        Ok(())
    }

    fn resolve_path(&self, rel_path: &str) -> AppResult<PathBuf> {
        let rel = Path::new(rel_path);
        if rel.is_absolute() || rel.components().any(|c| c == std::path::Component::ParentDir) {
            return Err(AppError::Other("invalid path".into()));
        }
        Ok(self.root.join(rel))
    }

    fn get_or_infer_meta(&self, rel_path: &str, content: &str) -> AppResult<KnowledgeMeta> {
        if let Ok(meta) = self.get_meta(rel_path) {
            return Ok(meta);
        }
        let title = infer_title(rel_path, content);
        let now = Utc::now();
        Ok(KnowledgeMeta {
            file_path: rel_path.to_string(),
            title,
            tags: vec![],
            created_at: now,
            updated_at: now,
        })
    }

    fn get_meta(&self, rel_path: &str) -> AppResult<KnowledgeMeta> {
        self.db.with_conn(|conn| {
            let row = conn.query_row(
                "SELECT file_path, title, created_at, updated_at FROM knowledge_meta WHERE file_path = ?1",
                params![rel_path],
                |row| {
                    Ok((
                        row.get::<_, String>(0)?,
                        row.get::<_, String>(1)?,
                        row.get::<_, String>(2)?,
                        row.get::<_, String>(3)?,
                    ))
                },
            )?;
            let tags = get_entity_tags(conn, "knowledge_tags", "file_path", rel_path)?;
            Ok(KnowledgeMeta {
                file_path: row.0,
                title: row.1,
                created_at: DateTime::parse_from_rfc3339(&row.2)
                    .map(|d| d.with_timezone(&Utc))
                    .unwrap_or_else(|_| Utc::now()),
                updated_at: DateTime::parse_from_rfc3339(&row.3)
                    .map(|d| d.with_timezone(&Utc))
                    .unwrap_or_else(|_| Utc::now()),
                tags,
            })
        })
    }
}

fn build_tree_node(root: &Path, current: &Path) -> KnowledgeTreeNode {
    let name = current
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("knowledge")
        .to_string();
    let rel = current
        .strip_prefix(root)
        .map(|p| p.to_string_lossy().replace('\\', "/"))
        .unwrap_or_default();

    let mut children = Vec::new();
    if let Ok(entries) = fs::read_dir(current) {
        let mut entries: Vec<_> = entries.filter_map(|e| e.ok()).collect();
        entries.sort_by_key(|e| e.file_name());
        for entry in entries {
            let path = entry.path();
            if path.is_dir() {
                children.push(build_tree_node(root, &path));
            } else if path.extension().map(|e| e == "md").unwrap_or(false) {
                children.push(KnowledgeTreeNode {
                    name: path.file_name().unwrap().to_string_lossy().to_string(),
                    path: path.strip_prefix(root).unwrap().to_string_lossy().replace('\\', "/"),
                    is_dir: false,
                    children: vec![],
                });
            }
        }
    }

    KnowledgeTreeNode {
        name,
        path: rel,
        is_dir: true,
        children,
    }
}

fn build_markdown(title: &str, tags: &[String], created: &DateTime<Utc>, body: &str) -> String {
    let tags_yaml = if tags.is_empty() {
        "  []".to_string()
    } else {
        tags.iter()
            .map(|t| format!("  - {t}"))
            .collect::<Vec<_>>()
            .join("\n")
    };
    let now = Utc::now();
    format!(
        "---\ntitle: {title}\ntags:\n{tags_yaml}\ncreated: {}\nupdated: {}\n---\n\n{body}",
        created.format("%Y-%m-%d"),
        now.format("%Y-%m-%d"),
    )
}

fn infer_title(rel_path: &str, content: &str) -> String {
    if let Some(line) = content.lines().find(|l| l.starts_with("# ")) {
        return line.trim_start_matches("# ").to_string();
    }
    Path::new(rel_path)
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("Untitled")
        .to_string()
}

fn sanitize_segment(s: &str) -> String {
    s.trim().trim_matches('/').to_string()
}

fn sanitize_filename(s: &str) -> String {
    s.chars()
        .map(|c| if "<>:\"/\\|?*".contains(c) { '_' } else { c })
        .collect::<String>()
        .replace(' ', "_")
}
